### PR TITLE
Data API V2: extending `NewServicesRepository` to build a complete list of services

### LINKED ISSUE #3277: Data API V2: Extending the functionality of `NewServicesRepository`
In the new data API package `data-api` the function `NewServicesRepository` is used to initialise a new object that defines where the data api definitions (the JSON files) are sitting for the server to load and process. 

This function needs to be extended in the following way:
- It should auto discover all subdirectories containing a `metadata.json`
- The `metadata.json` should be read to determine the `serviceType`, represented by `dataSource` in the file
- A unique map of `map[serviceName]FilePath` for each `serviceType` with validation to ensure a service doesn't have a duplicate definition e.g. a generated and a handwritten one - some of this functionality will need to be refactored out of `GetAll` which is where the paths are determined.

The directory structure is expected to look like the example below
```
├── README.md
├── handwritten-resource-manager
│   ├── AADB2C
│   │   ├── 2020-01-01
│   │   │   ├── ApiVersionDefinition.json
│   │   │   └── Example
│   │   │       ├── Constant-Example.json
│   │   │       ├── Model-ExampleResource.json
│   │   │       ├── Operation-Get.json
│   │   │       └── ResourceId-Example.json
│   │   └── ServiceDefinition.json
│   └── metadata.json
└── resource-manager
    ├── Advisor
    │   ├── 2020-01-01
    │   │   ├── ApiVersionDefinition.json
    │   │   └── Example
    │   │       ├── Constant-Example.json
    │   │       ├── Model-ExampleResource.json
    │   │       ├── Operation-Get.json
    │   │       └── ResourceId-Example.json
    │   └── ServiceDefinition.json
    └── metadata.json
```

NOTE: graph is not included yet, but will be added at a later date.

### LINKED ISSUE #3411: Data API V2: fix the caching
As described [in this comment](https://github.com/hashicorp/pandora/pull/3392#discussion_r1401995303) the Caching used in the Repository needs updating, at the time of writing:

* `GetAll` populates the cache in the event that it's unpopulated, and will return this.
* `GetByName` will _always_ load a Service from disk - but then (if the cache exists) will attempt to re-populate it.
* The locking needs updating to account for ^

Whilst [I should have caught this when reviewing this PR](https://github.com/hashicorp/pandora/pull/3151) - I missed this - however (based on the results below) this needs to be addressed prior to Data API v2 going live.

---

## Problem

The caching issues mean that currently it takes over 10 minutes for the Go SDK Generator to pull all the data from the Data API:

* First request: `2023/11/27 12:18:56`
* Last request: `2023/11/27 12:29:26`

Spinning up the Data API and hitting the `/v1/resource-manager/services` endpoint to warm the cache:

```
2023/11/27 12:17:41 "GET http://localhost:8080/v1/resource-manager/services HTTP/1.1" from [::1]:52231 - 200 12234B in 7.998263541s
2023/11/27 12:17:44 "GET http://localhost:8080/v1/resource-manager/services HTTP/1.1" from [::1]:52231 - 200 12234B in 361.958µs
2023/11/27 12:17:47 "GET http://localhost:8080/v1/resource-manager/services/Compute HTTP/1.1" from [::1]:52231 - 200 808B in 273.815875ms
```

and then launching `generator-go-sdk`, which starts making requests to the Data API ([full log](https://gist.github.com/tombuildsstuff/770bfbdde59ceed9b919204a971e2436)):

```
2023/11/27 12:18:56 "GET http://localhost:8080/v1/resource-manager/services HTTP/1.1" from [::1]:52250 - 200 12234B in 268.958µs
2023/11/27 12:18:56 "GET http://localhost:8080/v1/resource-manager/services/ManagedIdentity HTTP/1.1" from [::1]:52251 - 200 200B in 5.721708ms
2023/11/27 12:18:56 "GET http://localhost:8080/v1/resource-manager/services/Blueprints HTTP/1.1" from [::1]:52271 - 200 206B in 24.650083ms
2023/11/27 12:18:56 "GET http://localhost:8080/v1/resource-manager/services/PolicyInsights HTTP/1.1" from [::1]:52277 - 200 420B in 29.497458ms
2023/11/27 12:18:56 "GET http://localhost:8080/v1/resource-manager/services/TimeSeriesInsights HTTP/1.1" from [::1]:52268 - 200 206B in 32.136167ms
```

With final requests sent from `generator-go-sdk` to the Data API:

```
2023/11/27 12:29:24 "GET http://localhost:8080/v1/resource-manager/services/Network/2023-06-01/WebApplicationFirewallPolicies/schema HTTP/1.1" from [::1]:53479 - 200 219767B in 609.800042ms
2023/11/27 12:29:24 "GET http://localhost:8080/v1/resource-manager/services/Network/2023-06-01/ExpressRouteCrossConnectionRouteTable/operations HTTP/1.1" from [::1]:53480 - 200 359B in 587.02225ms
2023/11/27 12:29:25 "GET http://localhost:8080/v1/resource-manager/services/Network/2023-06-01/ExpressRouteCrossConnectionRouteTable/schema HTTP/1.1" from [::1]:53480 - 200 3112B in 585.039042ms
2023/11/27 12:29:25 "GET http://localhost:8080/v1/resource-manager/services/Network/2023-06-01/PrivateLinkService/operations HTTP/1.1" from [::1]:53480 - 200 356B in 581.870125ms
2023/11/27 12:29:26 "GET http://localhost:8080/v1/resource-manager/services/Network/2023-06-01/PrivateLinkService/schema HTTP/1.1" from [::1]:53480 - 200 112252B in 579.850875ms
```

However, skimming the log there's a number of cases where we take over 500ms to return e.g.:

```
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/ManagementGroups HTTP/1.1" from [::1]:52276 - 200 428B in 591.232417ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/FluidRelay HTTP/1.1" from [::1]:52348 - 200 297B in 507.343375ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/WebPubSub HTTP/1.1" from [::1]:52365 - 200 188B in 431.154583ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/ContainerInstance HTTP/1.1" from [::1]:52336 - 200 204B in 528.593042ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/Dynatrace HTTP/1.1" from [::1]:52311 - 200 294B in 559.319708ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/ApplicationInsights HTTP/1.1" from [::1]:52294 - 200 788B in 609.552292ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/VoiceServices HTTP/1.1" from [::1]:52337 - 200 306B in 563.322334ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/DeviceProvisioningServices HTTP/1.1" from [::1]:52356 - 200 345B in 553.103834ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/Maintenance HTTP/1.1" from [::1]:52287 - 200 424B in 655.832208ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/Cognitive HTTP/1.1" from [::1]:52282 - 200 188B in 668.088416ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/ChaosStudio HTTP/1.1" from [::1]:52368 - 200 316B in 472.375542ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/Search HTTP/1.1" from [::1]:52338 - 200 285B in 610.74575ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/ManagedApplications HTTP/1.1" from [::1]:52301 - 200 324B in 657.753917ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/Databricks HTTP/1.1" from [::1]:52328 - 200 543B in 650.351875ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/Communication HTTP/1.1" from [::1]:52377 - 200 306B in 489.94975ms
2023/11/27 12:18:57 "GET http://localhost:8080/v1/resource-manager/services/ElasticSan HTTP/1.1" from [::1]:52315 - 200 313B in 691.127042ms
```

---

The result of this is that `generator-go-sdk` takes far longer to run when using Data API v2 (~10m):

```
12:18PM tools/generator-go-sdk
 $ go build . && ./generator-go-sdk -output-dir=/path/to/github.com/hashicorp/go-azure-sdk --data-api=http://localhost:8080
2023/11/27 12:19:06 [DEBUG] Loading All Services..
2023/11/27 12:29:26 [DEBUG] Service "AnalysisServices"..
2023/11/27 12:29:26 [DEBUG] .. is opted out of generation, skipping..
2023/11/27 12:29:26 [DEBUG] Service "ServiceLinker"..
2023/11/27 12:29:26 [DEBUG] .. is opted out of generation, skipping..
2023/11/27 12:29:26 [DEBUG] Service "Elastic"..
2023/11/27 12:29:26 [DEBUG] .. is opted out of generation, skipping..
```

Compared to Data API v1 (which whilst takes far longer to build) - the cache means that requests are processed far quicker (~50s, [in GitHub Actions](https://github.com/hashicorp/pandora/actions/runs/7004230552/job/19051541789) - or 1m24s locally):

```
 1:50PM tools/generator-go-sdk
 $ go build . && ./generator-go-sdk -output-dir=/path/to/github.com/hashicorp/go-azure-sdk --data-api=http://localhost:5000
2023/11/27 13:50:38 [DEBUG] Loading All Services..
2023/11/27 13:52:02 [DEBUG] Service "ConnectedVmware"..
2023/11/27 13:52:02 [DEBUG] Service "ApplicationInsights"..
```

---

## Tasks

To fix this we need to:

1. Update `GetAll` to pull out a list of `Service Names` which are available on disk (and cache this).
2. Update both `GetAll` and `GetByName` to load the data from the cache if it's available, if not to load this from disk and then populate it.

Whilst this work would need to be done once #3392 has been merged.

## Spike

I've quickly spiked out the possible performance gains available - and by updating `GetByName` to load data from the cache if it doesn't exist - the average response time significantly decreases:

```
 $ make run
go build .
./data-api serve
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services HTTP/1.1" from [::1]:50234 - 200 12234B in 6.393124958s
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/DataLakeStore HTTP/1.1" from [::1]:50237 - 200 196B in 67.25µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/ServiceNetworking HTTP/1.1" from [::1]:50236 - 200 334B in 19.083µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/PostgreSqlHSC HTTP/1.1" from [::1]:50238 - 200 336B in 19.542µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/FluidRelay HTTP/1.1" from [::1]:50239 - 200 297B in 19.375µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/MariaDB HTTP/1.1" from [::1]:50251 - 200 408B in 20.417µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/Maps HTTP/1.1" from [::1]:50252 - 200 279B in 40.5µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/AlertsManagement HTTP/1.1" from [::1]:50256 - 200 557B in 22.292µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/MixedReality HTTP/1.1" from [::1]:50253 - 200 194B in 11.833µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/SignalR HTTP/1.1" from [::1]:50237 - 200 184B in 16.666µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/Communication HTTP/1.1" from [::1]:50236 - 200 306B in 17.792µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/EventHub HTTP/1.1" from [::1]:50258 - 200 412B in 26.458µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/Orbital HTTP/1.1" from [::1]:50254 - 200 184B in 36.666µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/ContainerRegistry HTTP/1.1" from [::1]:50259 - 200 708B in 19.542µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/LabServices HTTP/1.1" from [::1]:50257 - 200 424B in 15.875µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/RecoveryServicesBackup HTTP/1.1" from [::1]:50263 - 200 571B in 27.333µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/ApiManagement HTTP/1.1" from [::1]:50262 - 200 432B in 19.166µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/NewRelic HTTP/1.1" from [::1]:50260 - 200 307B in 16.25µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/Attestation HTTP/1.1" from [::1]:50255 - 200 300B in 14.416µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/Dashboard HTTP/1.1" from [::1]:50261 - 200 294B in 59.667µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/ConnectedVmware HTTP/1.1" from [::1]:50265 - 200 328B in 16.417µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/VoiceServices HTTP/1.1" from [::1]:50267 - 200 306B in 19.209µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/Elastic HTTP/1.1" from [::1]:50271 - 200 184B in 22.167µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/RedisEnterprise HTTP/1.1" from [::1]:50268 - 200 552B in 19.792µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/HDInsight HTTP/1.1" from [::1]:50281 - 200 310B in 15.541µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/DigitalTwins HTTP/1.1" from [::1]:50282 - 200 194B in 12.583µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/StorageCache HTTP/1.1" from [::1]:50264 - 200 303B in 13µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/NotificationHubs HTTP/1.1" from [::1]:50283 - 200 315B in 12.333µs
2023/11/27 12:57:01 "GET http://localhost:8080/v1/resource-manager/services/EventGrid HTTP/1.1" from [::1]:50266 - 200 188B in 12.541µs
```

For the sake of completeness / for comparison - I've compared the results of hitting the same Services across:

* Data API V1
* Data API V2 (without caching / as in `main` at `0dbebc24e282e06ae69457892cbc02dfe5dd662c`)
* Data API V2 (minimal caching spike - as outlined below)

| Service        | V1          | V2 (without caching) | V2 (minimal caching spike) |
|----------------|-------------|----------------------|----------------------------|
| API Management | `59.2710ms` | `2.144992458s`       | `19.166µs`                 | 
| Compute        | `3.2151ms`  | `2.239133708s`       | `68.958µs`                 | 
| EventGrid      | `14.7390ms` | `937.613834ms`       | `12.541µs`                 | 
| Maps           | `0.9006ms`  | `277.407875ms`       | `40.5µs`                   | 
| Network        | `8.3038ms`  | `3.192625833s`       | `154.959µs`                | 
| Storage        | `1.7315ms`  | `1.714739583s`       | `84.292µs`                 | 

The `minimal caching spike` used here is incomplete, only loading the data from the cache in `GetByName` if it already exists:

```diff
 func (s *ServicesRepositoryImpl) GetByName(serviceName string, serviceType ServiceType) (*ServiceDetails, error) {
+       if s.services != nil {
+               service, ok := (*s.services)[serviceName]
+               if ok {
+                       return &service, nil
+               }
+       }
+
        // GetByName builds the ServiceDetails for a singular service by calling processing functions to build the
        // structs for the ServiceApiVersionDetails and ServiceApiVersionResourceDetails
        serviceDirectory := fmt.Sprintf("%s/%s", s.directory, serviceName)
@@ -119,11 +130,9 @@ func (s *ServicesRepositoryImpl) GetByName(serviceName string, serviceType Servi
                return nil, fmt.Errorf("processing service definition for %s: %+v", serviceName, err)
        }

        if s.services != nil {
                s.Lock()
-               services := *s.services
-               services[serviceName] = *serviceDetails
+               (*s.services)[serviceName] = *serviceDetails
                s.Unlock()
        }

        return serviceDetails, nil
```

So we should see further gains when fully implemented.

### PR DESCRIPTION
**Pull request by stephybun**
Closes #3277
Closes #3411

**Summary:**
- `NewServicesRepository` has been updated to do the following:
    - Auto discovers subdirectories in the root dir of the api definitions with a `metadata.json`
    - Checks the DataSource specified in the `metadata.json` and builds a complete list of services for a given service type that populates the new attribute `serviceDirectory` which maps the service to a filepath
    - Validates that there aren't duplicate definitions for a service across several directories
    - Returns errors

- Updated the paths we use when loading the definitions to use the information stored in the new attribute `serviceDirectory`
- The entry point into the server/root routing handles the error for `NewServicesRepository` and exits if there is one
- `ServiceDefinition.json` is being loaded and processed now

TODO in a separate PR:
- Logging should be threaded through properly using whatever logging mechanism that chi relies on

### CODE REVIEW

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
On the whole this is looking good, but we should commit to either fixing or removing the cache, which (as discussed offline) I'll open an issue to track - however if the other comments can be addressed then this should otherwise be good to go 👍

File: tools/data-api/internal/repositories/services.go [hunk 1]
```diff
@@ -109,8 +166,7 @@ func (s *ServicesRepositoryImpl) GetAll(serviceType ServiceType) (*[]ServiceDeta
 func (s *ServicesRepositoryImpl) GetByName(serviceName string, serviceType ServiceType) (*ServiceDetails, error) {
 	// GetByName builds the ServiceDetails for a singular service by calling processing functions to build the
 	// structs for the ServiceApiVersionDetails and ServiceApiVersionResourceDetails
-	serviceDirectory := fmt.Sprintf("%s/%s", s.directory, serviceName)
-	if _, err := os.Stat(serviceDirectory); os.IsNotExist(err) {
+	if _, err := os.Stat(s.serviceDirectory[serviceName]); os.IsNotExist(err) {
```
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply tools/data-api/internal/repositories/services.go:169]: Opened https://github.com/hashicorp/pandora/issues/3411 to track this

**Review by tombuildsstuff — APPROVED** [review]
Aside from a handful of minor nits for consistency/styling this otherwise LGTM 👍

File: tools/data-api/internal/repositories/services.go [hunk 2]
```diff
@@ -270,7 +273,7 @@ func (s *ServicesRepositoryImpl) ProcessResourceDefinitions(serviceName string,
 	operations := make(map[string]ResourceOperations)
 	resourceIds := make(map[string]ResourceIdDefinition)
 
-	resourcePath := path.Join(s.serviceDirectory[serviceName], version, resource)
+	resourcePath := path.Join((*s.serviceNamesToDirectory)[serviceName], version, resource)
```
[the remark this answers is not in the document]
Comment (stephybun) [reply tools/data-api/internal/repositories/services.go:276]: `servicePath` is defined in `ProcessServiceDefinitions` where it's used multiple times.

This is a `ProcessResourceDefinitions` and it's only used once so I'm going to keep it as is.

File: tools/data-api/internal/repositories/services.go [hunk 3]
```diff
@@ -526,7 +529,7 @@ func (s *ServicesRepositoryImpl) ProcessTerraformDefinitions(serviceName string)
 		DataSources: make(map[string]TerraformDataSourceDetails),
 	}
 
-	terraformDefinitionsPath := path.Join(s.serviceDirectory[serviceName], "Terraform")
+	terraformDefinitionsPath := path.Join((*s.serviceNamesToDirectory)[serviceName], "Terraform")
```
[the remark this answers is not in the document]
Comment (stephybun) [reply tools/data-api/internal/repositories/services.go:532]: ditto

**Pull request merged**